src/evaluation.py

```diff
@@ -0,0 +1,68 @@
+from lstm_probing_model import lstm_probing
+import torchtext
+import torch
+import copy
+from torch import nn
+import numpy as np
+from torchtext.data.metrics import bleu_score
+import dataset
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+def collate_fn_decode(data):
+    hidden_states, explanation, explanation_2, explanation_3 = zip(*data)
+    return hidden_states, explanation, explanation_2, explanation_3
+
+def evaluate(tokenizer, model, encoder_states, exp_dataset, args):
+    training_dataset = dataset.CombinedDataset(encoder_states, exp_dataset, args.layer)
+    criterion = nn.CrossEntropyLoss()
+    return evaluate(tokenizer, model, training_dataset, criterion, args)
+
+def evaluate(tokenizer, model, dataset, criterion, args):
+    dataloader = DataLoader(dataset, batch_size=args.batch_size, collate_fn=collate_fn_decode)
+    model.eval()
+    overall_bleu = 0
+    decoded_words = []
+    reference_words = []
+    with torch.no_grad():
+        for batch, (x, y, y_2, y_3) in enumerate(tqdm(dataloader)):
+            x_tensor = torch.stack(list(x), dim=0).to(args.device)  # (batch, 768)
+            y_tensor = torch.stack(y, dim=0).to(args.device)  # (batch, 55)
+            for i in range(len(x_tensor)):
+                input_token = 1
+                hidden = model.init_hidden(x_tensor[i].unsqueeze(0))  # (batch, 768), (batch, 768)
+                hidden = (hidden[0].unsqueeze(0), hidden[1].unsqueeze(0))
+
+                sequence = [1]
+                for t in range(len(y[i])): 
+                    output, hidden, logits = model.generate(torch.tensor([[input_token]]).to(args.device), hidden)
+                    topv, topi = logits.topk(1)
+                    if topi.item() == 2: 
+                        break
+                    sequence.append(topi.item())
+                    input_token = topi.item()
+                sequence = sequence[1:]
+                ref_1 = copy.deepcopy(list(y[i]))
```

Comment:
I'm a bit confused as to why y is processed differently than y_2 and y_3. Is this because in training y is the source of truth, so it has been set up differently?